CPT: create Add Term component

timmyc opened the pull request:
This branch introduces the `<AddTerm/>` component which allows for creation of new terms for a given taxonomy:

__Default Display__
![new_project_ _testingtimmy2_wordpress_com_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/16813265/a504935c-48e5-11e6-9d92-cf3ca6766126.png)

__Dialog__
![new_project_ _testingtimmy2_wordpress_com_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/16813250/900babd4-48e5-11e6-9c32-4cf4542f97c0.png)

Currently you can add a new term using the dialog, much like the existing "Add Category" feature in the editor sidebar.

__To Test__
- Open up a portfolio project in the editor
- Expand the "Project Types" accordion
- Click the "Add New Project Type" button
- Interact with the modal form.  Use the search box to find a parent.
- Select a parent and note the "Top Level" checkbox unticks
- Play around with validation.  You should not be able to submit a term name that already exists for a given tree level ( i.e. only one top level term could be named `Major Key` ).
- Save a new term, note that it should appear in the original post-editor term selector.  If there are multiple pages of terms, you might need to scroll to trigger a new fetch, or simply search for your newly added term

__Todo in Future PR__
- Auto "checking" newly added term in post that is being edited


Test live: https://calypso.live/?branch=add/term-tree-selector/add-term

timmyc commented:
@aduth I was hoping to get some feedback/ideas on this PR from you.  Specifically - to perform the validation of the term name being added, should I just grab all the terms via `getTerms` and pass them in as a prop, or is there a more reduxy way to utilize the selector directly in the component code for a more targeted selector search?

Additionally, still kind of wondering the best approach to modify `terms` for the post that is being edited.  In the post-edit-store we currently listen for the `RECIEVE_ADD_TERM` action [[#](https://github.com/Automattic/wp-calypso/blob/master/client/lib/posts/post-edit-store.js#L394-L414)] - what are your thoughts here?

timmyc commented:
Added validation in 8c9736d.  One issue with this approach is if all terms have yet to be loaded ( if > 100 terms exist and user hasn't scrolled to load additional page(s) ) - the validation only runs against what is loaded locally.

aduth commented:
>Additionally, still kind of wondering the best approach to modify `terms` for the post that is being edited.

Haven't looked at the code yet, but maybe the reducer for edited posts can watch for the "add term success" action and if an edited post exists for that site, add it to its terms?

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+			siteId
+		};
+	},
+	( dispatch ) => {
```
You can pass a plain object as the second argument, which will have the same effect:

```js
{ addTerm }
```

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+import { addTerm } from 'state/terms/actions';
+
+class TermSelectorAddTerm extends Component {
+	static initialState = () => {
```
- Why does this need to be a function?
- Why do you need `Object.assign` ?

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	getSelectedValues() {
+		const termName = ReactDom.findDOMNode( this.refs.termName ).value.trim();
```
Should we track `termName` in state by changing the `<FormTextInput />` into a [controlled component](https://facebook.github.io/react/docs/forms.html#controlled-components) to avoid the `ref` and `findDOMNode`? Both work fine, but if for example we were to change `<FormTextInput />` to be wrapped in a div, this would no longer work.

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+		const termName = ReactDom.findDOMNode( this.refs.termName ).value.trim();
+		const parent = this.state.selectedParent.length ? this.state.selectedParent[ 0 ] : 0;
+
+		return {
```
With some variable renaming, this could be shortened to `return { name, parent };`

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+		const values = this.getSelectedValues();
+
+		if ( ! values.name.length ) {
+			error = true;
```
Rather than tracking `error`, why not just `return true` here and similarly `return` early in the conditions below, falling back to `return false;` at the end of the function.

aduth replied:
Oh, I see we set `error` into state. Was a bit thrown off by this being a boolean (where below it's an error message).

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+	( state, ownProps ) => {
+		const { taxonomy } = ownProps;
+		const siteId = getSelectedSiteId( state );
+		const taxonomyDetails = getPostTypeTaxonomy( state, siteId, ownProps.postType, taxonomy ) || {};
```
Can move `postType` into the destructuring a few lines above.

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+
+		return {
+			terms: getTerms( state, siteId, taxonomy ),
+			labels: taxonomyDetails.labels,
```
You could use shorthand here if you were to destructure the `getPostTypeTaxonomy`:

```js
const { taxonomy, postType } = ownProps;
const { labels } = getPostTypeTaxonomy( state, siteId, postType, taxonomy ) || {};
return { labels };
```

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+							ref="termName"
+							isError={ isError }
+							onKeyUp={ this.boundValidateInput } />
+						{ isError ? <FormInputValidation isError={ true } text={ this.state.error } /> : null }
```
As an FYI you can always omit `={ true }` e.g. `<FormInputValidation isError />`

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,232 @@
+	isValid() {
+		let error;
+
+		const values = this.getSelectedValues();
```
Should this be singular?

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -50,23 +54,29 @@ class EditorTermSelector extends Component {
+					siteId={ siteId }
+					multiple={ true }
+				/>
+				<AddTerm taxonomy={ taxonomyName } postType={ postType } />
```
Do we need `postType` to be passed from `<EditorTermSelector />`, or couldn't `<AddTerm />` do just the same lookup in its `connect`?

timmyc replied:
Potentially `<AddTerm />` could be used elsewhere, like the menu builder for example - where utilizing that selector will not work.  Right now in the menu builder I think we link back to wp-admin to add a new category.  Might be nice to drop this in there as well.

timmyc commented:
Okay @aduth I have fully reduxified this component now thanks to @nylen merging in #6548.  I would like to tackle the auto-selection of newly added terms in a follow-up PR as this one is already sitting at a hefty 300+ lines of new code.

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -26,21 +30,34 @@ class EditorTermSelector extends Component {
+		const { postTerms, taxonomyName, siteId, postId } = this.props;
 		const terms = cloneDeep( postTerms ) || {};
-		const taxonomyTerms = terms[ taxonomyName ] || {};
+		let taxonomyTerms = map( terms[ taxonomyName ] );
```
Why do we need `map` here?

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -26,21 +30,34 @@ class EditorTermSelector extends Component {
-		if ( taxonomyTerms[ selectedTerm.name ] ) {
-			delete taxonomyTerms[ selectedTerm.name ];
+				prev.push( term );
+				return prev;
```
Minor: This could be a one-liner push with `Array#concat` instead:

```js
return prev.concat( term );
```

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -26,21 +30,34 @@ class EditorTermSelector extends Component {
+			terms_by_id: {
+				[ taxonomyName ]: taxonomyTermIds.length ? taxonomyTermIds : null
+			}
+		}, siteId, postId );
```
Hmm, why do we need to assign both `terms` and `terms_by_id` when calling `editPost` ? So that they're kept in sync in the `getEditedPost` return value? Not sure I have a great alternative, but this seems to be quite the burden in using. I'm also curious how the REST API handles this case where we'd likely be passing both `terms` and `terms_by_id` in the request.

aduth replied:
Maybe the return value of `getEditedPost` could be normalized such that we don't include the full `terms` object, only `terms_by_id`. This avoids syncing concerns of both `terms` vs. `terms_by_id` and `terms` vs. `state.terms.items` (you'd need to call those selectors to get richer information about terms).

To that point, if we can't expect that the full terms are made available by the time we've received an existing post to be edited, maybe `state.terms.items` should be enhanced to pull the term objects from `RECEIVE_POST`? Don't recall if the structure of the objects is the same.

timmyc replied:
>  So that they're kept in sync in the getEditedPost return value

Correct.  This was being done prior by the `normalizer()` in the `post-edit-store`.

> Maybe the return value of getEditedPost could be normalized such that we don't include the full terms object, only terms_by_id

This is an option - but having the full object is currently used by the `<TermTokenField />` and am guessing will also be used to populate the selected labels in the taxonomy accordions.

We have had to do quite a bit of hackery to get terms to work properly here so I think the best long-term option is to strive for using the full `terms` object throughout, but change the shape of it to:

```
terms: {
    taxonomyName: [ { term }, { anotherTerm } ]
}
```

This will fix the problem with duplicate hierarchical term names, and the issue @nylen dealt with in `merge` yesterday.

The reason for leaning on `terms_by_id` here is the same as it was in #6468 - again the API not being able to handle the updating of `terms` if it is sent the same shape of an object that it responds with.

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -26,21 +30,34 @@ class EditorTermSelector extends Component {
+		}, siteId, postId );
 	}
 
 	getSelectedTermIds() {
```
If we have `terms_by_id` available to us, do we need `getSelectedTermIds` to map over `postTerms` ?

timmyc replied:
if we do go the route of `terms_by_id` being built-out in the normalizer then yes we could eliminate this - the API response for a post does not send `terms_by_id` - but supports it on `PUT` and `POST` requests

timmyc commented:
@aduth in 6e349fb - as we discussed yesterday I moved the `terms_by_id` logic into a new `getNormalizedPostEdits` selector.  Additionally I removed the old `terms_by_id` normalizer logic in the `post-edit-store`

aduth commented:
@timmyc Looks like my merge frenzy may have caused conflicts here.

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -26,21 +30,23 @@ class EditorTermSelector extends Component {
+		// map call transforms object returned by API into an array
+		let taxonomyTerms = map( terms[ taxonomyName ] ) || [];
+
+		if ( find( taxonomyTerms, { ID: selectedTerm.ID } ) ) {
```
To save a few cycles, maybe we should use a combination of `findIndex` and `Array#splice` instead of `find` and `reject`.

aduth commented on client/state/posts/selectors.js:
```diff
@@ -315,6 +316,23 @@ export function getPostEdits( state, siteId, postId ) {
+ * @param  {Number} postId Post ID
+ * @return {Object}        Normalized Post revisions
+ */
+export function getNormalizedPostEdits( state, siteId, postId ) {
```
Do we need to distinguish this from `getPostEdits`, or should we build normalization in there?

As background, we distinguish `getPost` from `getNormalizedPost` because edits need the raw post (from `getPost`) whereas display contexts want the normalized post (`getNormalizedPost`).

aduth commented on client/state/posts/utils.js:
```diff
@@ -91,3 +91,45 @@ export function mergeIgnoringArrays( object, ...sources ) {
+		return post;
+	}
+
+	post.terms_by_id = mapValues( taxonomies, ( taxonomy ) => {
```
I believe this is mutating the original post object.

In `getNormalizedPost`, we use `cloneDeep` to create an object clone, but that's mostly an artifact of `lib/post-normalizer` mutating the original objects. I'd say we should avoid this for these new normalizations and try to not mutate the object in the first place.

aduth commented on client/state/posts/utils.js:
```diff
@@ -91,3 +91,45 @@ export function mergeIgnoringArrays( object, ...sources ) {
+	taxonomies = reduce( taxonomies, ( prev, taxonomyTerms, taxonomyName ) => {
+		// Ensures we are working with an array
+		const termsArray = map( taxonomyTerms );
+		if ( termsArray && termsArray.length && ! isObject( termsArray[ 0 ] ) ) {
```
FYI, `isObject` is probably not what you wanted:

```
wp-calypso (add/term-tree-selector/add-term) $ node
> var isObject = require( 'lodash/isObject' );
undefined
> isObject( [] );
true
```

Instead, there's [`_.isPlainObject`](https://lodash.com/docs#isPlainObject)

aduth commented on client/state/posts/selectors.js:
```diff
@@ -315,6 +316,23 @@ export function getPostEdits( state, siteId, postId ) {
+ */
+export function getNormalizedPostEdits( state, siteId, postId ) {
+	const edits = getPostEdits( state, siteId, postId );
+	const normalize = flow( [
```
See #6715 for how I'm proposing to move normalization flows to `./utils.js` and let me know what you think.

timmyc commented:
@aduth rebased and applied feedback - including moving normalize flow and always normalizing edits - in 66ec396.  Small change in 93424c0 to add a compact class to the add button.

aduth commented on client/state/posts/utils.js:
```diff
@@ -91,3 +98,60 @@ export function mergeIgnoringArrays( object, ...sources ) {
+ * @param  {Ojbect} post Raw edited post object
+ * @return {Object}      Normalized post object
+ */
+export function normalizePost( post ) {
```
Should this be `normalizeEditedPost`? Since we treat posts distinctly between display and editing.

aduth replied:
Similarly `normalizeFlow` would be `normalizeEditedFlow`. End goal being have `normalizePost` for display and `normalizeEditedPost` for edited posts. Unless there's value in combining the two?

aduth commented on client/state/posts/utils.js:
```diff
@@ -91,3 +98,60 @@ export function mergeIgnoringArrays( object, ...sources ) {
+	if ( ! post || ! post.terms ) {
+		return post;
+	}
+	const normalizedPost = cloneDeep( post );
```
I think we could save on some performance by avoiding the `cloneDeep` here and using assign below, advantage of which being that only a shallow copy is created and we can return the original post (bypass the assign) when taxonomies is empty.

```js
return {
	...post,
	terms_by_id: mapValues( /* ... */ )
};
```

aduth commented on client/state/posts/utils.js:
```diff
@@ -91,3 +98,60 @@ export function mergeIgnoringArrays( object, ...sources ) {
+	// This can be detected by an array of strings vs an array of objects
+	taxonomies = reduce( taxonomies, ( prev, taxonomyTerms, taxonomyName ) => {
+		// Ensures we are working with an array
+		const termsArray = map( taxonomyTerms );
```
I already kinda forget how you'd explained this was working. In any case, does `toArray` achieve the same effect? https://lodash.com/docs#toArray

timmyc replied:
It is to ensure we are working with an array since the API returns an object for the post `terms`, toArray should work

aduth replied:
>It is to ensure we are working with an array since the API returns an object for the post `terms`, toArray should work

Yeah, I seem to recall you mentioning it was something to do with coercing into an array. I think `toArray` would be more clear to understand vs. `map`.

aduth commented:
When adding a term with a parent, I'm not seeing the new term appear in the accordion options until I either refresh or click "Add New" again. Seems to be fine for creating top-level terms.

timmyc commented:
> When adding a term with a parent, I'm not seeing the new term appear in the accordion options until I either refresh or click "Add New" again. Seems to be fine for creating top-level terms.

This should be fixed by 4396785

aduth commented on client/state/posts/test/utils.js:
```diff
@@ -141,4 +143,82 @@ describe( 'utils', () => {
+
+	describe( '#getTermIdsFromEdits()', () => {
+		it( 'should return the same post edit object if no term edits have been made', () => {
+			const normalizedPostEdits = getTermIdsFromEdits( {
```
Minor: Could improve "returns same" test to compare the references (`equal` vs. `eql` / `deep.equal`):

```
const original = { title: 'Chewbacca Saves' };
const normalizedPostEdits = getTermIdsFromEdits( original );

expect( normalizedPostEdits ).to.equal( original );
```

aduth commented:
>This should be fixed by 4396785

Hmm, doesn't seem the most robust (checking term length won't reflect changes to existing items, or if suddenly a term was dropped and another added in the same tick), but might be a good enough interim solution?

aduth commented on client/my-sites/term-tree-selector/terms.jsx:
```diff
@@ -94,7 +94,8 @@ const TermTreeSelectorList = React.createClass( {
 			prevProps.loading && ! this.props.loading ||
-			( ! prevProps.terms && this.props.terms )
+			( ! prevProps.terms && this.props.terms ) ||
+			( ( prevProps.terms && this.props.terms ) && ( prevProps.terms.length !== this.props.terms.length ) )
```
Aside: I've been using [`_.size`](https://lodash.com/docs#size) when wanting to get length when array is possibly undefined (in which case returns `0`, sufficient in some use cases including this one).

timmyc replied:
ah nice, i'll update to use `_.size`

aduth replied:
Would `prevProps.terms !== this.props.terms` be sufficient for checking that the terms has changed, since state _shouldn't_ be mutated?

aduth replied:
In `PostSelector`, [we call `recomputeRowHeights` when posts reference has changed](https://github.com/Automattic/wp-calypso/blob/6fabdf8aa234c901d44c620f52c6210d3d77d5af/client/my-sites/post-selector/selector.jsx#L121-L123), which calls `forceUpdate` under the hood ([docs](https://github.com/bvaughn/react-virtualized/blob/master/docs/VirtualScroll.md#recomputerowheights-index-number)). If effective, might be why we haven't encountered the same issue there.

timmyc commented:
> Hmm, doesn't seem the most robust (checking term length won't reflect changes to existing items, or if suddenly a term was dropped and another added in the same tick), but might be a good enough interim solution?

I think it is good enough for now, #6661 needs to be revisited and I could try for a better solution there.  Adding/Dropping in the same tick might be a difficult edge case to get into as-is right now.

aduth commented on client/my-sites/term-tree-selector/add-term.jsx:
```diff
@@ -0,0 +1,231 @@
+		}
+
+		const lowerCasedTermName = values.name.toLowerCase();
+		const matchingTerm = find( this.props.terms, ( term ) => {
```
No need for change, but [`Array#some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) / [`_.some`](http://lodash.com/docs#some) is more to the point of what your use-case is here (boolean return value if matching found, instead of full object).

aduth commented on client/my-sites/term-tree-selector/terms.jsx:
```diff
@@ -100,6 +102,10 @@ const TermTreeSelectorList = React.createClass( {
 		}
+
+		if ( this.props.terms !== prevProps.terms ) {
+			this.recomputeRowHeights();
```
No need to change here, but noticed in the `recomputeRowHeights` component function that we're calling both `recomputeRowHeights` and `forceUpdate` (the latter would be redundant and potentially affect performance).

aduth commented on client/post-editor/editor-term-selector/index.jsx:
```diff
@@ -26,21 +30,23 @@ class EditorTermSelector extends Component {
-		if ( taxonomyTerms[ selectedTerm.name ] ) {
-			delete taxonomyTerms[ selectedTerm.name ];
+		// map call transforms object returned by API into an array
+		const taxonomyTerms = map( terms[ taxonomyName ] ) || [];
```
Minor: Looks like both `_.map` and `_.toArray` are smart enough to result in an empty array when passed an undefined argument:

```
wp-calypso (add/term-tree-selector/add-term) $ node
> var map = require( 'lodash/map' );
undefined
> map( undefined );
[]
> var toArray = require( 'lodash/toArray' );
undefined
> toArray( undefined );
[]
```

aduth commented on client/state/posts/utils.js:
```diff
@@ -91,3 +98,60 @@ export function mergeIgnoringArrays( object, ...sources ) {
+/**
+ * Given a post object, returns a normalized post object
+ *
+ * @param  {Ojbect} post Raw edited post object
```
Typo `Ojbect` -> `Object` 😆

aduth commented on client/state/posts/selectors.js:
```diff
@@ -313,7 +314,7 @@ export function getEditedPost( state, siteId, postId ) {
 export function getPostEdits( state, siteId, postId ) {
 	const { edits } = state.posts;
-	return get( edits, [ siteId, postId || '' ], null );
+	return normalizeEditedPost( get( edits, [ siteId, postId || '' ], null ) );
```
Should we be applying the normalization in `getEditedPost` instead of here? I don't know that it makes much difference, but the `edits` object only contains revisions, not the full object.

aduth replied:
We may at some point (here or in a subsequent PR) want to memoize this function and/or the normalize function (noted similar follow-up task for post normalization in #6715).

aduth commented:
This is working well for me now. I left a few notes and some items I think we should seek to refine, either here or in a subsequent pull request, but I'd be fine with merging this as-is with iterations to come later.
